qiskit/extensions/hamiltonian_gate.py

```diff
@@ -0,0 +1,125 @@
+            # numpy matrix from `Operator.data`.
+            data = data.to_operator().data
+        # Convert to numpy array in case not already an array
+        data = numpy.array(data, dtype=complex)
+        # Check input is unitary
+        if not is_hermitian_matrix(data):
+            raise ExtensionError("Input matrix is not Hermitian.")
+        if isinstance(time, Number) and time != numpy.real(time):
+            raise ExtensionError("Evolution time is not real.")
+        # Check input is N-qubit matrix
+        input_dim, output_dim = data.shape
+        num_qubits = int(numpy.log2(input_dim))
+        if input_dim != output_dim or 2**num_qubits != input_dim:
+            raise ExtensionError(
+                "Input matrix is not an N-qubit operator.")
+
+        # Store instruction params
+        super().__init__('hamiltonian', num_qubits, [data, time], label=label)
+
+    def __eq__(self, other):
+        if not isinstance(other, HamiltonianGate):
+            return False
+        if self.label != other.label:
+            return False
+        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=False)
+        times_eq = self.params[1] == other.params[1]
+        return operators_eq and times_eq
+
+    def to_matrix(self):
+        """Return matrix for the unitary."""
+        try:
+            # pylint: disable=no-member
+            return scipy.linalg.expm(-1j * self.params[0] * float(self.params[1]))
+        except TypeError:
+            raise TypeError("Unable to generate Unitary matrix for "
+                            "unbound t parameter {}".format(self.params[1]))
+
+    def inverse(self):
+        """Return the adjoint of the unitary."""
+        return self.adjoint()
+
+    def conjugate(self):
+        """Return the conjugate of the Hamiltonian."""
+        return HamiltonianGate(numpy.conj(self.params[0]), -self.params[1])
+
+    def adjoint(self):
+        """Return the adjoint of the unitary."""
+        return HamiltonianGate(self.params[0], -self.params[1])
+
+    def transpose(self):
+        """Return the transpose of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), self.params[1])
+
+    def _define(self):
+        """Calculate a subcircuit that implements this unitary."""
+        q = QuantumRegister(self.num_qubits, 'q')
+        self.definition = [(UnitaryGate(self.to_matrix()), q[:], [])]
+
+    def qasm(self):
+        """Raise an error, as QASM is not defined for the HamiltonianGate."""
+        raise ExtensionError("HamiltonianGate as no QASM definition.")
+
+
+def hamiltonian(self, operator, time, qubits, label=None):
+    """Apply hamiltonian evolution to to qubits."""
+    if not isinstance(qubits, list):
+        qubits = [qubits]
+
+    return self.append(HamiltonianGate(data=operator, time=time, label=label), qubits, [])
+
+
+QuantumCircuit.hamiltonian = hamiltonian
```

Comment:
Do you want to add `HamiltonianGate` to `qiskit/extensions/__init__.py` so that `QuantumCircuit.hamiltonian` will be added on `import qiskit`? (Right now a user would have to explicitly call `from qiskit.extensions.hamiltonian_gate import HamiltonianGate` or similar before `qc.hamiltonian` will exist.)

Reply:
Done!